File: App/app/src/main/java/com/eth/vrcarnival/data/api/ApiService.kt
package com.eth.vrcarnival.data.api

import com.eth.vrcarnival.data.models.*
import retrofit2.Response
import retrofit2.http.*

interface ApiService {
    // Keep all existing working endpoints
    @POST("auth/send-otp")
    suspend fun sendOtp(@Body request: SendOtpRequest): Response<Unit>

    @POST("auth/verify-otp")
    suspend fun verifyOtp(@Body request: VerifyOtpRequest): Response<AuthResponse>

    @GET("v1/wallets/{address}/tokens")
    suspend fun getTokens(
        @Path("address") address: String,
        @Query("chainId") chainId: Int,
        @Query("limit") limit: Int = 20,
        @Query("page") page: Int = 1
    ): Response<ApiResponse<TokensResponse>>

    @GET("v1/wallets/{address}/balance")
    suspend fun getBalance(
        @Path("address") address: String,
        @Query("chainId") chainId: Int
    ): Response<ApiResponse<List<WalletBalance>>>

    @POST("v1/wallets/send")
    suspend fun sendTokens(
        @Header("Authorization") authorization: String,
        @Body request: SendTokenRequest
    ): Response<ApiResponse<SendTransactionResponse>>

    @POST("v1/verify-balance-change")
    suspend fun verifyBalanceChange(
        @Body request: VerifyBalanceChangeRequest
    ): Response<ApiResponse<Boolean>>

    @GET("v1/tokens")
    suspend fun listTokens(
        @Query("chainId") chainId: Int? = null,
        @Query("limit") limit: Int = 20,
        @Query("page") page: Int = 1,
        @Query("symbol") symbol: String? = null
    ): Response<ApiResponse<ListTokensResponse>>

    // KEEP: Working CAR token endpoints
    @GET("v1/wallets/{address}/token-balance")
    suspend fun getCarTokenBalance(
        @Path("address") address: String,
        @Query("chainId") chainId: Int
    ): Response<CarTokenBalance>

    @POST("v1/contracts/car-token/transfer")
    suspend fun transferCarTokens(
        @Header("Authorization") authorization: String,
        @Body request: TransferCarTokenRequest
    ): Response<ApiResponse<TransferCarTokenResponse>>

    // ADD: Unity integration endpoint
    @POST("ismobile")
    suspend fun sendWalletToUnity(@Body request: UnityIntegrationRequest): Response<Unit>

    // ADD: NFT endpoints to replace dummy data
    @GET("v1/contracts/{contractAddress}/universal-nfts")
    suspend fun getUniversalNFTs(
        @Path("contractAddress") contractAddress: String,
        @Query("walletAddress") walletAddress: String
    ): Response<NFTContractResponse>

    @GET("v1/contracts/{contractAddress}/erc1155/metadata/{tokenId}")
    suspend fun getNFTMetadata(
        @Path("contractAddress") contractAddress: String,
        @Path("tokenId") tokenId: String
    ): Response<NFTMetadataResponse>
}

File: VR Carnival/Assets/Scripts/APIService.cs
using System.Collections;
using System.Text;
using UnityEngine;
using UnityEngine.Networking;
using System.Threading.Tasks;
using System;
using System.Collections.Generic;
// ---------- AUTH ----------

[Serializable]
public class SendOtpRequest
{
    public string email;
}

[Serializable]
public class VerifyOtpRequest
{
    public string email;
    public string otp;
}

[Serializable]
public class AuthResponse
{
    public string token;
    public string refreshToken;
    public string walletAddress;
}

// ---------- TOKENS ----------

[Serializable]
public class SendTokenRequest
{
    public string toAddress;
    public string tokenAddress;
    public string amount;
    public int chainId;
}

[Serializable]
public class VerifyBalanceChangeRequest
{
    public string walletAddress;
    public string previousBalance;
}

[Serializable]
public class WalletBalance
{
    public string token;
    public string balance;
}

[Serializable]
public class TokensResponse
{
    public List<TokenData> tokens;
}

[Serializable]
public class TokenData
{
    public string name;
    public string symbol;
    public string balance;
}

// ---------- CAR TOKEN ----------

[Serializable]
public class CarTokenBalance
{
    public string tokenAddress;
    public string balance;
}

[Serializable]
public class TransferCarTokenRequest
{
    public string toAddress;
    public string amount;
}

[Serializable]
public class TransferCarTokenResponse
{
    public string txHash;
    public string status;
}

// ---------- UNITY INTEGRATION ----------

[Serializable]
public class UnityIntegrationRequest
{
    public string walletAddress;
    public string email;
}

// ---------- NFT ----------

[Serializable]
public class NFTContractResponse
{
    public List<NFTItem> nfts;
}

[Serializable]
public class NFTItem
{
    public string tokenId;
    public string name;
    public string image;
}

[Serializable]
public class NFTMetadataResponse
{
    public string name;
    public string description;
    public string image;
}

// ---------- GENERIC API WRAPPER ----------

[Serializable]
public class ApiResponse<T>
{
    public bool success;
    public string message;
    public T data;
}

public static class ApiService
{
    private static readonly string baseUrl = "http://81.15.150.175/";

    // ---------- AUTH ----------

    // public static async Task<UnityWebRequest> SendOtp(SendOtpRequest request)
    // {
    //     string url = baseUrl + "auth/send-otp";
    //     return await PostRequest(url, JsonUtility.ToJson(request));
    // }

    // public static async Task<UnityWebRequest> VerifyOtp(VerifyOtpRequest request)
    // {
    //     string url = baseUrl + "auth/verify-otp";
    //     return await PostRequest(url, JsonUtility.ToJson(request));
    // }

    // ---------- TOKENS & BALANCE ----------

    public static async Task<UnityWebRequest> GetTokens(string address, int chainId, int limit = 20, int page = 1)
    {
        string url = $"{baseUrl}v1/wallets/{address}/tokens?chainId={chainId}&limit={limit}&page={page}";
        return await GetRequest(url);
    }

    public static async Task<UnityWebRequest> GetBalance(string address, int chainId)
    {
        string url = $"{baseUrl}v1/wallets/{address}/balance?chainId={chainId}";
        return await GetRequest(url);
    }

    public static async Task<UnityWebRequest> SendTokens(string authToken, SendTokenRequest request)
    {
        string url = baseUrl + "v1/wallets/send";
        return await PostRequest(url, JsonUtility.ToJson(request), authToken);
    }

    public static async Task<UnityWebRequest> VerifyBalanceChange(VerifyBalanceChangeRequest request)
    {
        string url = baseUrl + "v1/verify-balance-change";
        return await PostRequest(url, JsonUtility.ToJson(request));
    }

    public static async Task<UnityWebRequest> ListTokens(int? chainId = null, int limit = 20, int page = 1, string symbol = null)
    {
        string query = $"limit={limit}&page={page}";
        if (chainId.HasValue) query += $"&chainId={chainId.Value}";
        if (!string.IsNullOrEmpty(symbol)) query += $"&symbol={symbol}";

        string url = baseUrl + "v1/tokens?" + query;
        return await GetRequest(url);
    }

    // ---------- CAR TOKEN ----------

    public static async Task<UnityWebRequest> GetCarTokenBalance(string address, int chainId)
    {
        string url = $"{baseUrl}v1/wallets/{address}/token-balance?chainId={chainId}";
        return await GetRequest(url);
    }

    public static async Task<UnityWebRequest> TransferCarTokens(string authToken, TransferCarTokenRequest request)
    {
        string url = baseUrl + "v1/contracts/car-token/transfer";
        return await PostRequest(url, JsonUtility.ToJson(request), authToken);
    }

    // ---------- UNITY INTEGRATION ----------

    public static async Task<UnityWebRequest> SendWalletToUnity(UnityIntegrationRequest request)
    {
        string url = baseUrl + "ismobile";
        return await PostRequest(url, JsonUtility.ToJson(request));
    }

    // ---------- NFT ENDPOINTS ----------

    public static async Task<UnityWebRequest> GetUniversalNFTs(string contractAddress, string walletAddress)
    {
        string url = $"{baseUrl}v1/contracts/{contractAddress}/universal-nfts?walletAddress={walletAddress}";
        return await GetRequest(url);
    }

    public static async Task<UnityWebRequest> GetNFTMetadata(string contractAddress, string tokenId)
    {
        string url = $"{baseUrl}v1/contracts/{contractAddress}/erc1155/metadata/{tokenId}";
        return await GetRequest(url);
    }

    // ====================== HELPERS ======================

    private static async Task<UnityWebRequest> PostRequest(string url, string jsonBody, string authToken = null)
    {
        UnityWebRequest request = new UnityWebRequest(url, "POST");
        byte[] bodyRaw = Encoding.UTF8.GetBytes(jsonBody);

        request.uploadHandler = new UploadHandlerRaw(bodyRaw);
        request.downloadHandler = new DownloadHandlerBuffer();
        request.SetRequestHeader("Content-Type", "application/json");
        if (!string.IsNullOrEmpty(authToken))
            request.SetRequestHeader("Authorization", authToken);

        var operation = request.SendWebRequest();
        while (!operation.isDone)
            await Task.Yield();

        Debug.Log($"POST {url} → {request.responseCode}");
        return request;
    }

    private static async Task<UnityWebRequest> GetRequest(string url, string authToken = null)
    {
        UnityWebRequest request = UnityWebRequest.Get(url);
        if (!string.IsNullOrEmpty(authToken))
            request.SetRequestHeader("Authorization", authToken);

        var operation = request.SendWebRequest();
        while (!operation.isDone)
            await Task.Yield();

        Debug.Log($"GET {url} → {request.responseCode}");
        return request;
    }
}
